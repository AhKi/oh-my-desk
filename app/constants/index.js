export const NUMBER_PER_PAGE = 10; // eslint-disable-line
