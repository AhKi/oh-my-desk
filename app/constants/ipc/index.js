export const WIDGET_SHOW_INACTIVE = 'WIDGET_SHOW_INACTIVE';
export const WIDGET_INFO_REQUEST = 'WIDGET_INFO_REQUEST';
export const WIDGET_INFO_RESULT = 'WIDGET_INFO_RESULT';
export const WIDGET_MANAGE = 'WIDGET_MANAGE';
export const WIDGET_OPEN = 'WIDGET_OPEN';
