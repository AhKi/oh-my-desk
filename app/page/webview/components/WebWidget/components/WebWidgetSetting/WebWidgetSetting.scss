.WebWidgetSetting {
  background-color: #ffffff;
  position: fixed;
}
