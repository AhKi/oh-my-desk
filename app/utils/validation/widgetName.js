function widgetName(value) {
  let validateMessage;

  if (value.length === 0) {
    validateMessage = 'Please enter the widget name.';
  } else {
    validateMessage = '';
  }

  return validateMessage;
}

export default widgetName;
