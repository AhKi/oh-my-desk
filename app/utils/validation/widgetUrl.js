import validUrl from 'valid-url';

function widgetUrl(value) {
  let validateMessage;

  if (value.length === 0) {
    validateMessage = 'Please enter the widget url.';
  } else if (!validUrl.isUri(value)) {
    validateMessage = 'Please match the URL format. (ex: https://www.google.com)';
  } else {
    validateMessage = '';
  }

  return validateMessage;
}

export default widgetUrl;
