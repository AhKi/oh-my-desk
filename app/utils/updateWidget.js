import * as IPC from 'constants/ipc';

function updateWidget(type, info) {
  window.ipcRenderer.send(IPC.WIDGET_MANAGE, {
    operation: 'UPDATE',
    widget: {
      id: info.id,
      type,
      name: info.name,
      url: info.url,
      position: info.position && {
        x: info.position.x,
        y: info.position.y,
      },
      size: info.size && {
        width: info.size.width,
        height: info.size.height,
      },
      transparency: info.transparency,
      isActive: info.isActive,
      isIcon: info.isIcon,
      isOnTop: info.isOnTop,
      favicon: info.favicon,
    },
  });
  window.ipcRenderer.send(IPC.WIDGET_INFO_REQUEST);
}

export default updateWidget;
