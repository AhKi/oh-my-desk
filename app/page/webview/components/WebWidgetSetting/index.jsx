import React from 'react';
import PropTypes from 'prop-types';
import ValidationInput from 'components/ValidationInput';
import OutsideClickHandler from 'components/OutsideClickHandler';
import validateName from 'utils/validation/widgetName';
import validateUrl from 'utils/validation/widgetUrl';
import './WebWidgetSetting.scss';

const propTypes = {
  name: PropTypes.string,
  widget: PropTypes.object, // eslint-disable-line
  url: PropTypes.string,
  onOpenPreference: PropTypes.func,
  onToggleSetting: PropTypes.func,
  onUpdateInfo: PropTypes.func,
};
const defaultProps = {
  name: '',
  widget: {},
  url: '',
  onOpenPreference() {},
  onToggleSetting() {},
  onUpdateInfo() {},
};

class WebWidgetSetting extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: props.name,
      nameError: '',
      url: props.url,
      urlError: '',
    };
    this.handleChangeName = this.handleChangeName.bind(this);
    this.handleChangeUrl = this.handleChangeUrl.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      name: nextProps.name,
      url: nextProps.url,
    });
  }

  handleChangeName(e) {
    this.setState({
      name: e.target.value,
      nameError: validateName(e.target.value),
    });
  }

  handleChangeUrl(e) {
    this.setState({
      url: e.target.value,
      urlError: validateUrl(e.target.value),
    });
  }

  handleSubmit(e) {
    e.preventDefault();
    const {
      name,
      nameError,
      url,
      urlError,
    } = this.state;
    const {
      widget,
      onToggleSetting,
      onUpdateInfo,
    } = this.props;

    if (name && url && (!nameError && !urlError)) {
      onUpdateInfo(widget.id, { name, url });
      onToggleSetting(false);
    }
  }

  render() {
    const {
      name,
      nameError,
      url,
      urlError,
    } = this.state;
    const { onOpenPreference, onToggleSetting } = this.props;

    return (
      <div className="WebWidgetSetting">
        <OutsideClickHandler onOutSideClick={onToggleSetting}>
          <form>
            <h6 className="WebWidgetSetting__title space-3x">
              <strong>
                Edit setting
              </strong>
            </h6>
            <ValidationInput
              autoFocus
              name="Name"
              value={name}
              error={nameError}
              onChange={this.handleChangeName}
            />
            <ValidationInput
              name="Url"
              value={url}
              error={urlError}
              onChange={this.handleChangeUrl}
            />
            <div className="WebWidgetSetting__btn-box">
              <button
                className="TextBtn TextBtn--blue WebWidgetSetting__text-btn"
                type="button"
                onClick={onOpenPreference}
              >
                Go to manage view
              </button>
              <input
                className="Btn Btn--primary Btn--sm"
                type="submit"
                value="Save"
                onClick={this.handleSubmit}
              />
            </div>
          </form>
        </OutsideClickHandler>
      </div>
    );
  }
}

WebWidgetSetting.propTypes = propTypes;
WebWidgetSetting.defaultProps = defaultProps;

export default WebWidgetSetting;
