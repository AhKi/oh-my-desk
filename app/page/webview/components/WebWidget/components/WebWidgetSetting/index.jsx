import React from 'react';
import PropTypes from 'prop-types';
import ValidationInput from 'components/ValidationInput';
import OutsideClickHandler from 'components/OutsideClickHandler';
import * as IPC from 'constants/ipc';
import validateName from 'utils/validation/widgetName';
import validateUrl from 'utils/validation/widgetUrl';
import updateWidget from 'utils/updateWidget';
import './WebWidgetSetting.scss';

const propTypes = {
  name: PropTypes.string,
  widget: PropTypes.object, // eslint-disable-line
  url: PropTypes.string,
  onToggleSetting: PropTypes.func,
};
const defaultProps = {
  name: '',
  widget: {},
  url: '',
  onToggleSetting() {},
};

class WebWidgetSetting extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: props.name,
      nameError: '',
      url: props.url,
      urlError: '',
    };
    this.handleChangeName = this.handleChangeName.bind(this);
    this.handleChangeUrl = this.handleChangeUrl.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.openManageView = this.openManageView.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      name: nextProps.name,
      url: nextProps.url,
    });
  }

  openManageView() { // eslint-disable-line
    window.ipcRenderer.send(IPC.WIDGET_MANAGER_OPEN);
  }

  handleChangeName(e) {
    this.setState({
      name: e.target.value,
      nameError: validateName(e.target.value),
    });
  }

  handleChangeUrl(e) {
    this.setState({
      url: e.target.value,
      urlError: validateUrl(e.target.value),
    });
  }

  handleSubmit(e) {
    e.preventDefault();
    const {
      name,
      nameError,
      url,
      urlError,
    } = this.state;

    if (name && url && (!nameError && !urlError)) {
      this.props.onToggleSetting(false);
      updateWidget(
        'web',
        Object.assign({}, this.props.widget, {
          name: this.state.name,
          url: this.state.url,
        }),
      );
    }
  }

  render() {
    const {
      name,
      nameError,
      url,
      urlError,
    } = this.state;
    const { onToggleSetting } = this.props;

    return (
      <div className="WebWidgetSetting">
        <OutsideClickHandler onOutSideClick={onToggleSetting}>
          <form>
            <h3>Edit setting</h3>
            <ValidationInput
              autoFocus
              name="Name"
              value={name}
              error={nameError}
              onChange={this.handleChangeName}
            />
            <ValidationInput
              name="Url"
              value={url}
              error={urlError}
              onChange={this.handleChangeUrl}
            />
            <button
              type="button"
              onClick={this.openManageView}
            >
              Go to manage view
            </button>
            <input
              type="submit"
              value="Save"
              onClick={this.handleSubmit}
            />
          </form>
        </OutsideClickHandler>
      </div>
    );
  }
}

WebWidgetSetting.propTypes = propTypes;
WebWidgetSetting.defaultProps = defaultProps;

export default WebWidgetSetting;
