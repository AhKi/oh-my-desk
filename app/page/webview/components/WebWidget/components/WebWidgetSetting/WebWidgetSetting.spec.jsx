import React from 'react';
import { shallow } from 'enzyme';

import * as IPC from 'constants/ipc';
import * as utils from 'utils/updateWidget';
import * as widgetName from 'utils/validation/widgetName';
import * as widgetUrl from 'utils/validation/widgetUrl';
import WebWidgetSetting from '.';

describe('<WebWidgetSetting />', () => {
  it('should match to snapshot when render default', () => {
    const wrapper = shallow(<WebWidgetSetting />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should change state when change props', () => {
    const componentWillReceiveProps = jest.spyOn(WebWidgetSetting.prototype, 'componentWillReceiveProps');
    const wrapper = shallow(
      <WebWidgetSetting
        name="initial-name"
        url="initial-url"
      />,
    );

    wrapper.setProps({
      name: 'mock-name',
      url: 'mock-url',
    });
    expect(componentWillReceiveProps).toHaveBeenCalledTimes(1);
    expect(wrapper.state().name).toBe('mock-name');
    expect(wrapper.state().url).toBe('mock-url');
  });

  it('should call window.ipcRenderer.send when call openManageView', () => {
    const ipcRenderer = {
      send() {},
    };
    const send = jest.spyOn(ipcRenderer, 'send');
    window.ipcRenderer = ipcRenderer;
    const wrapper = shallow(<WebWidgetSetting />);

    wrapper.instance().openManageView();

    expect(send).toHaveBeenCalledTimes(1);
    expect(send).toHaveBeenCalledWith(IPC.WIDGET_MANAGER_OPEN);
  });

  it('should change state when call handleChangeName', () => {
    const validationName = jest
      .spyOn(widgetName, 'default')
      .mockImplementation(() => 'mock-validation-name');
    const wrapper = shallow(<WebWidgetSetting />);

    wrapper.instance().handleChangeName({
      target: {
        value: 'mock-name',
      },
    });

    expect(wrapper.state().name).toBe('mock-name');
    expect(validationName).toHaveBeenCalledTimes(1);
    expect(validationName).toHaveBeenCalledWith('mock-name');
    expect(wrapper.state().nameError).toBe('mock-validation-name');
  });

  it('should change state when call handleChangeName', () => {
    const validationUrl = jest
      .spyOn(widgetUrl, 'default')
      .mockImplementation(() => 'mock-validation-url');
    const wrapper = shallow(<WebWidgetSetting />);

    wrapper.instance().handleChangeUrl({
      target: {
        value: 'mock-name',
      },
    });

    expect(wrapper.state().url).toBe('mock-name');
    expect(validationUrl).toHaveBeenCalledTimes(1);
    expect(validationUrl).toHaveBeenCalledWith('mock-name');
    expect(wrapper.state().urlError).toBe('mock-validation-url');
  });

  describe('test handleSubmit', () => {
    const event = {
      preventDefault() {},
    };
    const onToggleSetting = jest.fn();
    const updateWidget = jest.spyOn(utils, 'default');
    const wrapper = shallow(
      <WebWidgetSetting
        onToggleSetting={onToggleSetting}
      />,
    );

    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('when name, url is matched', () => {
      wrapper.setState({
        name: 'mock-name',
        nameError: '',
        url: 'https://google.com',
        urlError: '',
      });

      wrapper.instance().handleSubmit(event);

      expect(onToggleSetting).toHaveBeenCalledTimes(1);
      expect(onToggleSetting).toHaveBeenCalledWith(false);
      expect(updateWidget).toHaveBeenCalledTimes(1);
      expect(updateWidget).toHaveBeenCalledWith(
        'web',
        {
          name: 'mock-name',
          url: 'https://google.com',
        },
      );
    });

    it('when name, url is not matched', () => {
      wrapper.setState({
        name: 'mock-name',
        nameError: 'mock-name-error',
        url: 'https://google.com',
        urlError: 'mock-url-error',
      });

      wrapper.instance().handleSubmit(event);

      expect(onToggleSetting).toHaveBeenCalledTimes(0);
      expect(updateWidget).toHaveBeenCalledTimes(0);
    });
  });
});
