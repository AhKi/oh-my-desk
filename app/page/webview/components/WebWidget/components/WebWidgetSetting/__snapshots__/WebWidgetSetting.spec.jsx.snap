// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WebWidgetSetting /> should match to snapshot when render default 1`] = `
<div
  className="WebWidgetSetting"
>
  <OutsideClickHandler
    onOutSideClick={[Function]}
  >
    <form>
      <h3>
        Edit setting
      </h3>
      <ValidationInput
        autoFocus={true}
        className="ValidationInput"
        error=""
        errorClassName="ValidationInput__validation-message"
        inputClassName="ValidationInput__text-input"
        name="Name"
        nameClassName="ValidationInput__label"
        onChange={[Function]}
        placeholder=""
        value=""
      />
      <ValidationInput
        autoFocus={false}
        className="ValidationInput"
        error=""
        errorClassName="ValidationInput__validation-message"
        inputClassName="ValidationInput__text-input"
        name="Url"
        nameClassName="ValidationInput__label"
        onChange={[Function]}
        placeholder=""
        value=""
      />
      <button
        onClick={[Function]}
        type="button"
      >
        Go to manage view
      </button>
      <input
        onClick={[Function]}
        type="submit"
        value="Save"
      />
    </form>
  </OutsideClickHandler>
</div>
`;
